lib/src/Pos/Worker/Block.hs

```diff
@@ -146,27 +149,139 @@ blockCreator
+                                   finalHeavyPsk
+    if | weAreLeader || weAreHeavyDelegate ->
+            onNewSlotWhenLeader genesisConfig txpConfig slotId pske diffusion
+        | otherwise -> pass
+
+  where
+    logOnEpochFS = if siSlot == localSlotIndexMinBound then logInfoS else logDebugS
+    logOnEpochF = if siSlot == localSlotIndexMinBound then logInfo else logDebug
+
+dropObftEbb ::
+       forall ctx m.
+       ( BlockWorkMode ctx m
+       )
+    => Genesis.Config
+    -> TxpConfiguration
+    -> m ()
+dropObftEbb genesisConfig txpConfig = do
+    -- not sure if everything needs to run inside StateLock
+    tipHeader <- DB.getTipHeader
+    logDebug $ sformat ("dropObftEbb: tipHeader: ("%shown%").") tipHeader
+    case tipHeader of
+        BlockHeaderMain _      -> pure ()
+        BlockHeaderGenesis _   -> do
+            mbEbbBlund <- getBlund (configGenesisHash genesisConfig)
+                                   (blockHeaderHash tipHeader)
+            case mbEbbBlund of
+                Nothing -> Prelude.error "Pos.Worker.Block.dropObftEbb: unable to get blund for EBB"
+                Just ebbBlund -> do
+                    rollbackBlocks genesisConfig $ NewestFirst (ebbBlund :| [])
+                    normalizeMempool genesisConfig txpConfig
+
+blockCreatorOriginal
+    :: ( BlockWorkMode ctx m
+       , HasMisbehaviorMetrics ctx
+       )
+    => Genesis.Config
+    -> TxpConfiguration
+    -> SlotId
+    -> Diffusion m -> m ()
+blockCreatorOriginal genesisConfig txpConfig (slotId@SlotId {..}) diffusion = do
     -- First of all we create genesis block if necessary.
     mGenBlock <- createGenesisBlockAndApply genesisConfig txpConfig siEpoch
-    whenJust mGenBlock $ \createdBlk -> do
-        logInfo $ sformat ("Created genesis block:\n" %build) createdBlk
-        jsonLog $ jlCreatedBlock (configEpochSlots genesisConfig) (Left createdBlk)
-
-    -- Then we get leaders for current epoch.
-    leadersMaybe <- LrcDB.getLeadersForEpoch siEpoch
-    case leadersMaybe of
-        -- If we don't know leaders, we can't do anything.
-        Nothing -> logWarning "Leaders are not known for new slot"
-        -- If we know leaders, we check whether we are leader and
-        -- create a new block if we are. We also create block if we
-        -- have suitable PSK.
-        Just leaders ->
-            maybe onNoLeader
-                  (onKnownLeader leaders)
-                  (leaders ^? ix (fromIntegral $ getSlotIndex siSlot))
+
+    -- the ConsensusEra could've changed due to to this
+    -- call trace (thing above calls thing below):
+
+    --    blockCreatorOriginal
+    -- -> createGenesisBlockAndApply
+    -- -> createGenesisBlockDo
+    -- -> verifyBlocksPrefix
+    -- -> usVerifyBlocks
+    -- -> verifyBlock
+    -- -> processGenesisBlock
+    -- -> adoptBlockVersion
+
+    era <- getConsensusEra
+    case era of
+        OBFT _ -> blockCreator genesisConfig txpConfig slotId diffusion
```

Comment:
I wrote this bit, so I'm suggesting this to myself 🙂 but I think some explanation could help:

```suggestion
        -- If the ConsensusEra has changed due to `createGenesisBlockAndApply`
        -- being run, then we want to switch over to `blockCreatorObft`, while
        -- running the logic which precedes its call in `blockCreator`.
        -- So we just re-enter `blockCreator`.
        
        OBFT _ -> blockCreator genesisConfig txpConfig slotId diffusion
```